### PR TITLE
Enhance setup.py cmake errors reporting (#111)

### PR DESCRIPTION
**Pull request by caliuf**
Handle possible cmake errors by displaying them to the user.

Referenced in #111

### CODE REVIEW

**Comment by McSinyx** [comment]
Thank you for submitting this patch!  Since I'm a bit lazy, could you please show us the output of `pip install .` when an error occurs.  A screenshot might be preferable if it fits in one screen, mainly for me to see the colors.

I'll review this somewhere within tomorrow.  Until then, please add your name to the copyright header of the file you're modifying and fix the linting errors raised by flake8.

**Comment by caliuf** [comment]
Here's how it looks:
setup_error

I'll print on the stderr instead of stdout by the way (but this doesn't change output colors in my system)

**Comment by caliuf** [comment]
I updated the patch. With the most recent pip version the output is the same in the screenshot, aside that it's all red.

Check if I changed copyright notice correctly.

File: setup.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 # setup script
-# Copyright (C) 2019, 2020  Nguyễn Gia Phong
+# Copyright (C) 2019, 2020  Nguyễn Gia Phong, Francesco Caliumi
```
Comment (Huy-Ngo) [inline setup.py]: We have copyright notice for each author on a separate line. Yours would look like this:

```python
# Copyright (C) 2020  Francesco Caliumi
```

File: setup.py [hunk 2]
```diff
@@ -57,8 +58,19 @@ def finalize_options(self) -> None:
...
+            cmake = run(['cmake', '.'],
+                        check=True, stdout=DEVNULL, stderr=PIPE,
+                        cwd=self.build_temp, universal_newlines=True)
+        except CalledProcessError as e:
+            print(f"\n\ncmake: Error occurred!\n\n"
+                  f"(Standard output)\n"
+                  f"{e.stdout}\n\n"
+                  f"(Standard error)\n"
+                  f"{e.stderr}\n",
+                  file=sys.stderr)
```
Comment (Huy-Ngo) [inline setup.py]: In this project, we prefer to use single quotation mark `'` to double quotation mark `"`

**Review by Huy-Ngo — CHANGES_REQUESTED** [review]
Thank you for submitting this patch. I just have a few comments on the copyright notice and code style.

File: setup.py (hunk 2 above)
Comment (McSinyx) [inline setup.py]: The following seems to be sufficient:
```suggestion
            print(e.stderr, file=sys.stderr)
```

File: setup.py [hunk 3]
```diff
@@ -57,8 +58,19 @@ def finalize_options(self) -> None:
...
+            sys.exit(1)
+
```
Comment (McSinyx) [inline setup.py]: I'd prefer a re-raise here, since it'd contain the error message which has the exit code of CMake.  
```suggestion
            raise
```

**Review by McSinyx — CHANGES_REQUESTED** [review]

File: setup.py (hunk 2 above)
Comment (caliuf) [reply setup.py]: @McSinyx I'm unfamiliar with cmake and if stdout could help to clarify what's going on if an error arise, so it's ok for me. 

Giving no clues about what the cmake error output is could create confusion in a distracted user, let me now if you are sure about this

File: setup.py (hunk 3 above)
Comment (caliuf) [reply setup.py]: I did this because the stacktrace is very misleading on what the error really is, but I see your point I'll change that

**Comment by caliuf** [comment]
I updated the PR with your suggestions. This is what it looks like now (pip 20.2, python 3.7):

setup-error

File: setup.py (hunk 2 above)
Comment (McSinyx) [reply setup.py]: As shown in your screenshot, CMake actually prints `CMake Error at ...` when there's a problem and such output will always be delivered to stderr.  In addition the script to be executed here also prints to stderr, so there's never any output to stdout.

File: setup.py (hunk 3 above)
Comment (McSinyx) [reply setup.py]: You got a point about the traceback, so how about `raise e from None`?

File: setup.py [hunk 4]
```diff
@@ -17,6 +18,7 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with palace.  If not, see <https://www.gnu.org/licenses/>.
 
+import sys
```
Comment (McSinyx) [inline setup.py:21]: flake8 is complaining about this

**Comment by caliuf** [comment]
One of the appveyor.com jobs died for a timeout, I think it's mergeable

**Review by Huy-Ngo — APPROVED** [review]
Looks good to me

**Comment by McSinyx** [comment]
@caliuf, did you see my comment at https://github.com/McSinyx/palace/pull/112#discussion_r462889554?

**Comment by caliuf** [comment]
@McSinyx I didn't know the `raise X from Y` syntax, and I can hardly find any mention of it on the web, what does it do exactly? 
At first glance it seems that Y should be an (exception) object, a form of "X error happened while processing Y", but I can't see how it can help in this case.

I've tried it and the output seems exactly the same

**Review by McSinyx — APPROVED** [review]
You're right, [raise x from y](https://docs.python.org/3/library/exceptions.html) doesn't help the traceback if there's only one exception raised through the handling process.  I'll go ahead merging this now and iterate on it later.  Thanks for your patience, @caliuf!

**Pull request merged**